File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/floors/DungeonSection.java
package com.joeshuff.dddungeongenerator.generator.floors;

import android.graphics.Point;
import com.joeshuff.dddungeongenerator.generator.dungeon.Dungeon;
import com.joeshuff.dddungeongenerator.generator.dungeon.Room;
import com.joeshuff.dddungeongenerator.generator.generating.DelauneyTriangulate;
import com.joeshuff.dddungeongenerator.generator.generating.MinSpanningTree;
import com.joeshuff.dddungeongenerator.generator.generating.PathFinding;
import com.joeshuff.dddungeongenerator.generator.models.Rectangle;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class DungeonSection  {

    private int id;

    private int height = 0;
    private int width = 0;

    private transient Dungeon mainDungeon;
    private transient Floor floor;

    private Point startPoint = new Point();

    private transient List<DelauneyTriangulate.Triangle> triangulateGraph = new ArrayList<>();
    private transient List<MinSpanningTree.Edge> minSpanningTree = new ArrayList<>();
    private transient List<MinSpanningTree.Edge> triangularEdges = new ArrayList<>();

    private List<List<Point>> corridors = new ArrayList<>();
    private transient List<Point> centers = new ArrayList<>();
    private transient List<MinSpanningTree.Edge> finalConnections = new ArrayList<>();

    public List<Room> rooms = new ArrayList<>();

    public static boolean linearProgression = false;

    public DungeonSection(int id, Dungeon mainDungeon, Floor floor, int width, int height, Point p) {
        this(id, mainDungeon, floor, width, height);
        this.startPoint = p;
    }

    public DungeonSection(int id, Dungeon mainDungeon, Floor floor, int width, int height) {
        this.id = id;
        this.mainDungeon = mainDungeon;
        this.floor = floor;
        this.height = height;
        this.width = width;
    }

    public int getId() {
        return id;
    }

    public Point getStartPoint() {
        return startPoint;
    }

    protected void branchOut() {
        int coverage = 0;
        int totalCoverage = height * width;

        while (coverage < mainDungeon.getGlobalModifier().getMapCoveragePercentage() * totalCoverage) {
            coverage = 0;

            for (Room r : rooms) {
                coverage += r.getArea() * 1.5f;
            }

            rooms.add(new Room(this, rooms.size()  + 1, mainDungeon.getRnd(), mainDungeon.getGlobalModifier()));
        }
    }

    public void calculateNearestPartner() {
        boolean allDone = false;

        while (!allDone) {
            allDone = true;

            for (Room r1 : rooms) {

                if (r1.isRejected()) continue;

                Room nearest = null;
                int largestCoverage = 0;

                for (Room r2 : rooms) {
                    if (r1 == r2) continue;
                    if (r2.isRejected()) continue;

                    int coverage = r1.intersects(r2);
                    if (coverage > largestCoverage) {
                        nearest = r2;
                        largestCoverage = coverage;
                    }
                }

                if (nearest != null) {
                    if (r1.moveAwayFrom(nearest)) {
                        allDone = false;
//                        System.out.println("Nearest to " + r1.getId() + " is " + nearest.getId());
                    }
                }
            }
        }
    }

    public void calculateRejectedRooms() {
        int validRooms = 0;

        for (Room r : rooms) {
            r.calculateRejected(rooms);
            if (r.isSelected() && !r.isRejected()) validRooms ++;
        }

        if (validRooms == 0) {
            Room best = null;
            double fail = Integer.MAX_VALUE;

            for (Room r : rooms) {
                if (r.calcHowMuchItFailedBy() < fail) {
                    fail = r.calcHowMuchItFailedBy();
                    best = r;
                }
            }

            best.forceSelect();
        }
    }

    public void triangulate() {
        centers = new ArrayList<>();

        for (Room r : rooms) {
            if (!r.isSelected() || r.isRejected()) continue;
            centers.add(r.myCenter());
        }

        triangulateGraph = DelauneyTriangulate.calculate(centers);
    }

    public void minSpanningTree() {
        if (triangulateGraph.isEmpty()) return;

        triangularEdges = new ArrayList<>();

        for (DelauneyTriangulate.Triangle triangle : triangulateGraph) {
            triangularEdges.addAll(triangle.getEdges());
        }

        List<MinSpanningTree.Edge> tempEdges = new ArrayList<>();
        tempEdges.addAll(triangularEdges);

        minSpanningTree = MinSpanningTree.calculate(tempEdges, centers);
    }

    public void combinePaths() {
        finalConnections = new ArrayList<>();
        finalConnections.addAll(minSpanningTree);

        if (linearProgression) return;

        for (MinSpanningTree.Edge edge : triangularEdges) {
            if (finalConnections.contains(edge)) continue;

            if (mainDungeon.getRnd().nextDouble() <= mainDungeon.getGlobalModifier().getTriangulationAdditionChange()) {
                finalConnections.add(edge);
            }
        }
    }

    public void pathFind() {
        corridors = PathFinding.findPaths(this, rooms, finalConnections);
    }

    public void finalise() {
        finalConnections.clear();
        minSpanningTree.clear();
    }

    public void clearUnnecessaryData() {
        triangularEdges.clear();
        triangulateGraph.clear();
        rooms.removeIf(room -> !room.isSelected() || room.isRejected());
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public Floor getFloor() {
        return floor;
    }

    public List<Room> getRooms() {
        return rooms;
    }

    public List<List<Point>> getCorridors() {
        return corridors;
    }

    public List<List<Point>> getGlobalCorridors() {
        List<List<Point>> globalCorridors = new ArrayList<>();

        for (List<Point> localCorridor : getCorridors()) {

            List<Point> globalCorridor = new ArrayList<>();

            for (Point point : localCorridor) {
                globalCorridor.add(new Point(point.x + startPoint.x, point.y + startPoint.y));
            }

            globalCorridors.add(globalCorridor);
        }

        return globalCorridors;
    }

    public List<DelauneyTriangulate.Triangle> getTriangulateGraph() {
        return triangulateGraph;
    }

    public List<DelauneyTriangulate.Triangle> getGlobalTriangulateGraph() {
        List<DelauneyTriangulate.Triangle> globalTriangles = new ArrayList<>();

        for (DelauneyTriangulate.Triangle triangle : getTriangulateGraph()) {
            globalTriangles.add(triangle.translate(startPoint));
        }

        return globalTriangles;
    }

    public List<MinSpanningTree.Edge> getGlobalFinalConnections() {
        List<MinSpanningTree.Edge> globalEdges = new ArrayList<>();

        for (MinSpanningTree.Edge edge : getFinalConnections()) {
            globalEdges.add(edge.translate(startPoint));
        }

        return globalEdges;
    }

    public List<MinSpanningTree.Edge> getFinalConnections() {
        return finalConnections;
    }

    public List<MinSpanningTree.Edge> getGlobalMinSpanningTree() {
        List<MinSpanningTree.Edge> globalEdges = new ArrayList<>();

        for (MinSpanningTree.Edge edge : getMinSpanningTree()) {
            globalEdges.add(edge.translate(startPoint));
        }

        return globalEdges;
    }

    public List<MinSpanningTree.Edge> getMinSpanningTree() {
        return minSpanningTree;
    }

    public Dungeon getMainDungeon() {
        return mainDungeon;
    }

    public Rectangle me() {
        return new Rectangle(startPoint.x, startPoint.y, width, height);
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof DungeonSection)) return false;
        return ((DungeonSection) o).id == id;
    }
}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/floors/DungeonSection.kt
package com.joeshuff.dddungeongenerator.generator.floors

import com.joeshuff.dddungeongenerator.db.models.Point
import com.joeshuff.dddungeongenerator.db.toRealmList
import com.joeshuff.dddungeongenerator.generator.dungeon.Dungeon
import com.joeshuff.dddungeongenerator.generator.dungeon.Room
import com.joeshuff.dddungeongenerator.generator.generating.DelauneyTriangulate
import com.joeshuff.dddungeongenerator.generator.generating.DelauneyTriangulate.Triangle
import com.joeshuff.dddungeongenerator.generator.generating.MinSpanningTree
import com.joeshuff.dddungeongenerator.generator.generating.PathFinding
import com.joeshuff.dddungeongenerator.generator.models.Corridor
import com.joeshuff.dddungeongenerator.generator.models.Rectangle
import com.joeshuff.dddungeongenerator.util.Logs
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.Ignore
import java.lang.Exception
import java.security.spec.ECField

open class DungeonSection(
        var id: Int,
        @Ignore val floor: Floor?,
        var width: Int,
        var height: Int): RealmObject() {

    var startPoint: Point? = Point()
        private set

    @Ignore
    private var triangulateGraph: ArrayList<Triangle> = ArrayList()

    @Ignore
    private var minSpanningTree: ArrayList<MinSpanningTree.Edge> = ArrayList()

    @Ignore
    private var triangularEdges: ArrayList<MinSpanningTree.Edge> = ArrayList()

    var corridors: RealmList<Corridor> = RealmList()
        private set

    @Ignore
    private var centers: ArrayList<Point> = ArrayList()

    @Ignore
    private var finalConnections: ArrayList<MinSpanningTree.Edge> = ArrayList()

    var rooms: RealmList<Room> = RealmList()

    constructor(): this(0, null, 0, 0)

    constructor(id: Int, floor: Floor, width: Int, height: Int, p: Point) : this(id, floor, width, height) {
        startPoint = p
    }

    fun branchOut() {
        var coverage = 0
        val totalCoverage = height * width
        val gennedRooms = arrayListOf<Room>()

        try {
            floor?.dungeon?.let {
                while (coverage < it.getGlobalModifier().getMapCoveragePercentage() * totalCoverage) {
                    coverage = gennedRooms.map { (it.area * 1.5f).toInt() }.sum()
                    Logs.i("COVERAGE", "$coverage out of ${it.getGlobalModifier().getMapCoveragePercentage() * totalCoverage}")

                    gennedRooms.add(Room(rooms.size + 1, this, it.rnd, it.getGlobalModifier()))
                }
            }
        } catch (e: Exception) {
            throw e
        }

        rooms = gennedRooms.toRealmList()
    }

    fun calculateNearestPartner() {
        var allDone = false

        while (!allDone) {
            allDone = true

            for (r1 in rooms) {

                if (r1.isRejected) continue

                var nearest: Room? = null

                var largestCoverage = 0

                for (r2 in rooms) {
                    if (r1 == r2) continue
                    if (r2.isRejected) continue

                    val coverage = r1.intersects(r2)
                    if (coverage > largestCoverage) {
                        nearest = r2
                        largestCoverage = coverage
                    }
                }

                if (nearest != null) {
                    if (r1.moveAwayFrom(nearest)) {
                        allDone = false
//                        Logs.i("Nearest to " + r1.getId() + " is " + nearest.getId());
                    }
                }
            }
        }
    }

    fun calculateRejectedRooms() {
        var validRooms = 0

        for (r in rooms) {
            r.calculateRejected(rooms)
            if (r.isSelected && !r.isRejected) validRooms++
        }

        if (validRooms == 0) {
            var best: Room? = null
            var fail = Int.MAX_VALUE.toDouble()

            for (r in rooms) {
                if (r.calcHowMuchItFailedBy() < fail) {
                    fail = r.calcHowMuchItFailedBy()
                    best = r
                }
            }

            best?.forceSelect()
        }
    }

    fun triangulate() {
        centers = ArrayList()

        for (r in rooms) {
            if (!r.isSelected || r.isRejected) continue
            centers.add(r.myCenter())
        }

        triangulateGraph = DelauneyTriangulate.calculate(centers)
    }

    fun minSpanningTree() {
        if (triangulateGraph.isEmpty()) return

        triangularEdges = ArrayList()

        for (triangle in triangulateGraph) {
            triangularEdges.addAll(triangle.edges)
        }

        val tempEdges: ArrayList<MinSpanningTree.Edge> = ArrayList()
        tempEdges.addAll(triangularEdges)

        minSpanningTree = MinSpanningTree.calculate(tempEdges, centers)
    }

    fun combinePaths() {
        floor?.dungeon?.let {
            finalConnections = ArrayList()
            finalConnections.addAll(minSpanningTree)

            if (linearProgression) return

            for (edge in triangularEdges) {
                if (finalConnections.contains(edge)) continue

                if (it.rnd.nextDouble() <= it.getGlobalModifier().getTriangulationAdditionChange()) {
                    finalConnections.add(edge)
                }
            }
        }
    }

    fun pathFind() {
        corridors.clear()
        PathFinding(this, rooms, finalConnections).findPaths().forEach {
            corridors.add(it)
        }
    }

    fun finalise() {
        finalConnections.clear()
        minSpanningTree.clear()
    }

    fun complete() {
        rooms.forEach { it.complete() }
    }

    fun clearUnnecessaryData() {
        triangularEdges.clear()
        triangulateGraph.clear()
        rooms.removeAll { room: Room -> !room.isSelected || room.isRejected }
    }

    fun getRooms(): List<Room> {
        return rooms
    }

    fun getGlobalCorridors(): List<Corridor> {
        val globalCorridors: ArrayList<Corridor> = ArrayList()

        for (localCorridor in corridors) {
            globalCorridors.add(localCorridor.globalise(startPoint?.x?: 0, startPoint?.y?: 0))
        }

        return globalCorridors
    }

    fun getTriangulateGraph(): List<Triangle> {
        return triangulateGraph
    }

    fun getGlobalTriangulateGraph(): List<Triangle> {
        val globalTriangles: ArrayList<Triangle> = ArrayList()

        for (triangle in getTriangulateGraph()) {
            startPoint?.let {
                globalTriangles.add(triangle.translate(it))
            }
        }

        return globalTriangles
    }

    fun getGlobalFinalConnections(): List<MinSpanningTree.Edge> {
        val globalEdges: ArrayList<MinSpanningTree.Edge> = ArrayList()

        for (edge in getFinalConnections()) {
            startPoint?.let {
                globalEdges.add(edge.translate(it))
            }
        }

        return globalEdges
    }

    fun getFinalConnections(): List<MinSpanningTree.Edge> {
        return finalConnections
    }

    fun getGlobalMinSpanningTree(): List<MinSpanningTree.Edge> {
        val globalEdges: ArrayList<MinSpanningTree.Edge> = ArrayList()

        for (edge in getMinSpanningTree()) {
            startPoint?.let {
                globalEdges.add(edge.translate(it))
            }
        }

        return globalEdges
    }

    fun getMinSpanningTree(): List<MinSpanningTree.Edge> {
        return minSpanningTree
    }

    fun me(): Rectangle? {
        startPoint?.let {
            return Rectangle(it.x, it.y, width, height)
        }

        return null
    }

    override fun equals(o: Any?): Boolean {
        return if (o !is DungeonSection) false else o.id == id
    }

    companion object {
        @Ignore
        var linearProgression = false
    }
}